### ISSUE TITLE
Segmentation fault (after using -d command?)

### ISSUE THREAD

**Issue by svatonop** [biocore/sortmerna#200]
Kindly asking for any idea or/and help. I've got a 'Segmentation fault' after running sortmerna command bellow. The last edit in my code was defining -d path (I was previously using sortmerna version 2.1). In log it is terminated with:
```
main:55 Running task ALIGN_REPORT: 4
Readstats::calculate starting ...   /root/sortmerna-3.0.3/src/sortmerna/gzip.cpp:120  WARNING: inflateEnd status is 0
```
I am using version 3.0.3 and command:
`sortmerna --ref ./Indices/tRNAs.fasta,./Indices/tRNAs:/home/svatonp/local/bin/sortmerna/rRNA_databases/silva-euk-18s-id95.fasta,./Indices/silva-euk-18s:/home/svatonp/local/bin/sortmerna/rRNA_databases/silva-euk-28s-id98.fasta,./Indices/silva-euk-28s:/home/svatonp/local/bin/sortmerna/rRNA_databases/rfam-5s-database-id98.fasta,./Indices/rfam-5s:/home/svatonp/local/bin/sortmerna/rRNA_databases/rfam-5.8s-database-id98.fasta,./Indices/rfam-5.8s --reads-gz ./reads_folder/reads.fastq.gz --aligned ./03_SortMeRNA/reads.null --other ./03_SortMeRNA/reads --fastx -d ./Indices/for_sortmerna/ -a 7 -e 1 --log -v &>./Indices/temp.log
`
Note: before each run I delete and create for_sortmerna folder.

Appreciate any help...

**Comment by biocodz** [comment]
Please, use full paths for files everywhere with any option i.e. no `./..`'
Side notes:
Version 3.0.4 is better to use than 3.0.3
Version 4.0 to be released next week allows for relative paths and simplifies the options format.

**Comment by svatonop** [comment]
I have replaced all relative paths with the full paths but still the same error (in terminal: Segmentation fault, in log:
`main:55 Running task ALIGN_REPORT: 4
Readstats::calculate starting ...   /root/sortmerna-3.0.3/src/sortmerna/gzip.cpp:120  WARNING: inflateEnd status is 0`

**Comment by svatonop** [comment]
This time I havent cleaned folder for -d algorithm, the log contains:
`Fast CRC32 supported: Not supported on x86`
Could be the bug?

**Comment by biocodz** [comment]
what is the output of `uname -a` on your machine?

**Comment by svatonop** [comment]
`Linux 3.10.0-693.17.1.el7.x86_64.debug #1 SMP Thu Jan 25 04:40:07 CST 2018 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by biocodz** [comment]
Seems your system is Centos/RHEL. What's the output of the following
```
cat /etc/centos-release
rpm -q centos-release
lsb_release -d
```
Also, would be good to see the complete command you use, and all the trace it outputs.

**Comment by svatonop** [comment]
I am working on computing machine of my institute, so that is all managed by IT department. OS:
`NAME="Scientific Linux"
VERSION="7.1 (Nitrogen)"`

**Comment by svatonop** [comment]
Basically I run the command above, input reads and outputs are stored to remote storage, indexes are in my home (local/bin/sortmerna) of the computing machine.

I can request for reinstallation of sortmerna and try again. Could I mention to pay attention on some dependences? RocksDB?

**Comment by biocodz** [comment]
I would advise to install the version 3.0.4, if it gets to the re-installation.
3.0.4 distro only has a single file for Linux [sortmerna-3.0.4-Linux.sh](https://github.com/biocore/sortmerna/releases/download/v3.0.4/sortmerna-3.0.4-Linux.sh), and is statically linked, meaning all the dependencies are already packaged with the binaries (no worries about rocksdb).
`Scientific Linux` is a RHEL based distro, and 3.0.4 installer was created on Centos/RHEL, so has better chances to work on your machine.
As I understand, you work in a networked environment, with remote storage etc. This certainly complicates the picture for me.
If you look e.g. in the [issue 198](https://github.com/biocore/sortmerna/issues/198), you'll see the correct steps and the complete execution trace (scroll down the message list). I need all that info when troubleshooting.

**Comment by svatonop** [comment]
seems I proceed... Now I am working with the version 3.0.4, but I received the same 'Segmentation fault' as previous. I've just remember, that for the first run I hadn't known about setting of Key-value DB location and it was created without problem, but next run it did complain about the non-empty folder. So then I specified path to remote storage where I was managing to create folder before run of command and remove after. And that started causing the 'Segmentation fault'. Now it seems to run OK. I am using the path to computing machine for kvdb creation. The warning `[inflatez:120] WARNING: inflateEnd status is 0` is still present, but log is now increasing and no error appears on terminal (still running):

`[main:55] Running task ALIGN_REPORT: 4

[calculate:52] Starting ...   [inflatez:120] WARNING: inflateEnd status is 0
[calculate:147] Done. Elapsed time: 160.64 sec. Reads processed: 46263196

[align:381] Using number of Processor threads set in run options: 7
Number of cores: 16 Read threads:  1 Write threads: 1 Processor threads: 7
[ThreadPool:36] initialized Pool with: [9] threads

[ReadsQueue:58] read_queue created
[ReadsQueue:58] write_queue created

[Refstats:32] Index Statistics calculation Start ... Done. Time elapsed: 9.53 sec
[align:414] Loading index 0 part 1/1 ... done [0.83] sec
[align:425] Loading references  ... done [0.20] sec

[write:19] Writer writer_0 thread 139809416259328 started
[run:40] Processor proc_0 thread 139809424652032 started
[run:40] Processor proc_1 thread 139809449830144 started
[run:40] Processor proc_4 thread 139809407866624 started
[run:40] Processor proc_3 thread 139809441437440 started
[run:40] Processor proc_2 thread 139809399473920 started
[run:40] Processor proc_5 thread 139809382688512 started
[run:40] Processor proc_6 thread 139809391081216 started
[read:43] reader_0thread: 139809433044736 started
[pop:105] read_queue Popped id:                99999
[pop:105] write_queue Popped id:                99999
[pop:105] read_queue Popped id:               199999
[pop:105] write_queue Popped id:               199999
[pop:105] read_queue Popped id:               299999
....`

**Comment by biocodz** [comment]
You have to empty the `kvdb/` prior Each run of Sortmerna.

Using `-d` to set the location for DB is Optional. If you don't specify it, the program will use the default location `$HOME/sortmerna/run/kvdb/`. Each time you start Sortmerna, it will check the `kvdb/` is empty, just to make sure it won't delete some valuable user data. If the `kvdb/` doesn't have any data of yours (and it shouldn't. Only data internal to Sortmerna), you can delete that directory without hesitation i.e. `rm -rf $HOME/sortmerna/run/kvdb/`, or whatever path Sortmerna prompts you. 

The `WARNING: inflateEnd status is 0` can be safely ignored.

**Comment by biocodz** [comment]
The ultimate reason for this problem was the failure of the code to create the work directory. It failed silently without generating an error, and then tried to initiate the key-value DB, which caused the Segfault.
This problem is now fixed in [Release 4.0](https://github.com/biocore/sortmerna/releases/tag/v4.0.0).
We encourage you to use the 4.0 version, as it implements many improvements including native paired reads processing.
Closing this for now...
